File: src/domain/models/hand.ts
import { Tile } from '../../utils/tileParser';
import { Mentsu } from './mentsu';
import { WaitType } from './waitType';

export class HandDecomposition {
  readonly mentsuList: Mentsu[];
  readonly jantai: Tile[]; // 雀頭 (2枚)
  readonly waitType: WaitType;

  constructor(mentsuList: Mentsu[], jantai: Tile[], waitType: WaitType) {
    this.mentsuList = mentsuList;
    this.jantai = jantai;
    this.waitType = waitType;
  }

  get isChitoitsu(): boolean { return false; }
  get isKokushi(): boolean { return false; }
}

export class ChitoitsuDecomposition extends HandDecomposition {
  readonly pairs: Tile[][];

  constructor(pairs: Tile[][]) {
    super([], [], WaitType.tanki);
    this.pairs = pairs;
  }

  override get isChitoitsu(): boolean { return true; }
}

export class KokushiDecomposition extends HandDecomposition {
  constructor() {
    super([], [], WaitType.tanki);
  }

  override get isKokushi(): boolean { return true; }
}

export interface Hand {
  tiles: Tile[];
  winTile: Tile;
  isTsumo: boolean;
  isMenzen: boolean;
  isParent: boolean;
  isRiichi: boolean;
  isIppatsu: boolean;
  dora: Tile[];
  seatWind?: Tile;
  roundWind?: Tile;
  openMentsu: Mentsu[];
}

export function createHand(params: {
  tiles: Tile[];
  winTile: Tile;
  isTsumo?: boolean;
  isMenzen?: boolean;
  isParent?: boolean;
  isRiichi?: boolean;
  isIppatsu?: boolean;
  dora?: Tile[];
  seatWind?: Tile;
  roundWind?: Tile;
  openMentsu?: Mentsu[];
}): Hand {
  return {
    tiles: params.tiles,
    winTile: params.winTile,
    isTsumo: params.isTsumo ?? false,
    isMenzen: params.isMenzen ?? true,
    isParent: params.isParent ?? false,
    isRiichi: params.isRiichi ?? false,
    isIppatsu: params.isIppatsu ?? false,
    dora: params.dora ?? [],
    seatWind: params.seatWind,
    roundWind: params.roundWind,
    openMentsu: params.openMentsu ?? [],
  };
}

File: lib/domain/models/hand.dart
import '../../utils/tile_parser.dart';
import 'mentsu.dart';
import 'wait_type.dart';

class HandDecomposition {
  final List<Mentsu> mentsuList;
  final List<Tile> jantai; // 雀頭 (2枚)
  final WaitType waitType;

  const HandDecomposition({
    required this.mentsuList,
    required this.jantai,
    required this.waitType,
  });

  bool get isChitoitsu => false;
  bool get isKokushi => false;
}

class ChitoitsuDecomposition extends HandDecomposition {
  final List<List<Tile>> pairs;

  ChitoitsuDecomposition({required this.pairs})
      : super(
          mentsuList: const [],
          jantai: const [],
          waitType: WaitType.tanki,
        );

  @override
  bool get isChitoitsu => true;
}

class KokushiDecomposition extends HandDecomposition {
  KokushiDecomposition()
      : super(
          mentsuList: const [],
          jantai: const [],
          waitType: WaitType.tanki,
        );

  @override
  bool get isKokushi => true;
}

class Hand {
  final List<Tile> tiles;
  final Tile winTile;
  final bool isTsumo;
  final bool isMenzen;
  final bool isParent;
  final bool isRiichi;
  final bool isIppatsu;
  final List<Tile> dora;
  final Tile? seatWind;
  final Tile? roundWind;
  final List<Mentsu> openMentsu;

  const Hand({
    required this.tiles,
    required this.winTile,
    this.isTsumo = false,
    this.isMenzen = true,
    this.isParent = false,
    this.isRiichi = false,
    this.isIppatsu = false,
    this.dora = const [],
    this.seatWind,
    this.roundWind,
    this.openMentsu = const [],
  });
}
